### ISSUE TITLE
Trim operations for selection don't always work properly

### ISSUE THREAD

**Issue by ArkadiuszMichalski** [notepad-plus-plus/notepad-plus-plus#12658]
Trim leading operation for first partially selected line not work.

For such example with one line selection:
![image](https://user-images.githubusercontent.com/2730894/207974161-b0967638-f444-4b85-a336-c436256a05f3.png)
Trim Trailing works
![image](https://user-images.githubusercontent.com/2730894/207974372-7a1c082c-3130-416d-84b6-b065ad5de7e6.png)
but Trim Leading not 
![image](https://user-images.githubusercontent.com/2730894/207974423-6c625b39-ff96-4818-8cb7-bf486129f795.png)

For multiple lines it's the same:
```
		ABCD
		ABCD		
		ABCD		
ABCD		
```
![image](https://user-images.githubusercontent.com/2730894/208005459-38621ef4-66f8-4def-8cf4-28fedd462636.png)

Run for this second case `Trim Leading and Trailing Space` command. This first tab on first line (covered by selection) is not removed, but last tab on fourth line (covered by selection) is removed. It's a bit confusing that deleting in selection doesn't work the same way.

![image](https://user-images.githubusercontent.com/2730894/208202507-be31a330-a7fd-4184-be22-e97629027f7d.png)

```
Notepad++ v8.4.7   (32-bit)
Build time : Dec  1 2022 - 02:39:50
Path : D:\_test\NPPClean\notepad++.exe
Command Line : -multiInst
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
OS Name : Windows 11 (64-bit) 
OS Version : 21H2
OS Build : 22000.1219
Current ANSI codepage : 1250
Plugins : 
    mimeTools (2.8)
    NppConverter (4.4)
    NppExport (0.4)

```

**Comment by ArkadiuszMichalski** [comment]
There is also one another problem. For belowe case:
![image](https://user-images.githubusercontent.com/2730894/208203461-923a3232-cd78-4f3a-86b4-d193de127aed.png)
and run `Trim Leading and Trailing Space` command we get:
![image](https://user-images.githubusercontent.com/2730894/208203595-677957e8-d801-492d-8f2a-e2e137dc01d0.png)
but should be:
![image](https://user-images.githubusercontent.com/2730894/208203634-9fff0f20-9047-4332-8880-44bf96a9307c.png)
Selection cover only 2 tabs and only this tabs should be removed.

This is because the first call to `doTrim()` removes 2 tabs and the selection is empty, so the second call to `doTrim()` thinks there was no selection and operates on the entire document.

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f6d36235cd1dbcbfdccada385c32b241aae426e6/PowerEditor/src/NppCommands.cpp#L1900-L1909

I fix this also in https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12655.

**Comment by ArkadiuszMichalski** [comment]
@donho 

Can you still take a look at it? When changing https://github.com/notepad-plus-plus/notepad-plus-plus/commit/bf34ef0dd568b792627813a4445adf649a5c5ba6 I didn't notice that `doTrim()` is used in other commands.

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f6d36235cd1dbcbfdccada385c32b241aae426e6/PowerEditor/src/NppCommands.cpp#L1918-L1929

Now, when we do selection, `doTrim()` will operate on the selection, but the rest above code on the entire document. I need to correct the above command as well, but first I want to fix `doTrim()`.

I will probably make a new `eol2ws()` function and use it in the places below, it will be easier to change it in future.

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f6d36235cd1dbcbfdccada385c32b241aae426e6/PowerEditor/src/NppCommands.cpp#L1911
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f6d36235cd1dbcbfdccada385c32b241aae426e6/PowerEditor/src/NppCommands.cpp#L1918

**Comment by donho** [comment]
@ArkadiuszMichalski 

> I will probably make a new `eol2ws()` function and use it in the places below, it will be easier to change it in future.

Hmm... I don't feel about adding stuffs and stuffs to make a complicate concept work. 

As I mentioned here:
https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12655#issuecomment-1356365695

Trim operations are rather belong to Line Operation, so let's keep this logic and code both simple.

**Comment by alankilborn** [comment]
>Trim operations rather belong to Line Operations, so let's keep this logic and code both simple.

I would think to emphasize that, the "trim" choices should be moved under the _Line Operations_ submenu?

**Comment by donho** [comment]
@alankilborn 
> I would think to emphasize that, the "trim" choices should be moved under the _Line Operations_ submenu?

It won't be moved under Line Operation sub-menu for 3 reasons:
1. They treat the blank characters TAB and White Space. 
2. They've been where they are since long time ago.
3. There are never request from users to change their behaviour or their category.

So I would say just let's keep their current behaviour and their position.

**Comment by ArkadiuszMichalski** [comment]
@donho
>I will probably make a new eol2ws() function and use it in the places below, it will be easier to change it in future.
>>Hmm... I don't feel about adding stuffs and stuffs to make a complicated concept work.
>>As I mentioned here https://github.com/notepad-plus-plus/notepad-plus-plus/pull/12655#issuecomment-1356365695, Trim operations rather belong to Line Operations, so let's keep this logic and code both simple.

But `eol2ws` will not to be for `doTrim()`. I meant another fix. This belowe two place:
 https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f6d36235cd1dbcbfdccada385c32b241aae426e6/PowerEditor/src/NppCommands.cpp#L1911
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f6d36235cd1dbcbfdccada385c32b241aae426e6/PowerEditor/src/NppCommands.cpp#L1918

use the same logic:
```
			_pEditView->execute(SCI_TARGETWHOLEDOCUMENT);
			_pEditView->execute(SCI_LINESJOIN);
```

Second `IDM_EDIT_TRIMALL` also use `doTrim()` at the begining so I need fix `doTrim()` (mainly determine the behavior of the selection, because `IDM_EDIT_TRIMALL` and `IDM_EDIT_EOL2WS` must/should behave the same).

By creating a new `eol2ws` I will cover both `IDM_EDIT_EOL2WS` and `IDM_EDIT_TRIMALL` (here's the second part). I mention this because by introducing selection for `doTrim()`, those two places above should also be fixed (especially `IDM_EDIT_TRIMALL` because it uses `doTrim()` so now it just doesn't work right).

Let's focus on `doTrim()` first.

**Comment by donho** [comment]
@ArkadiuszMichalski 

OK, I see the limitation of method of `doTrim()`.
From the dev perspective, I understand your viewpoint and method. But I should also be in the shoes of users.

> But eol2ws will not to be for doTrim(). I meant another fix. This belowe two place:
> case IDM_EDIT_EOL2WS: 
> case IDM_EDIT_TRIMALL: 
> use the same logic:

The refactoring can always come after, there's no urgency for doing new eol2ws() method, IMO. 

> By creating a new eol2ws I will cover both IDM_EDIT_EOL2WS and IDM_EDIT_TRIMALL (here's the second part). I mention this because by introducing selection for doTrim(), those two places above should also be fixed (especially IDM_EDIT_TRIMALL because it uses doTrim() so now it just doesn't work right).

Firstly, AFAIK, there's no bug report regarding to Trim operations, so there's nothing to be fixed in these operations.
Secondly, we can improve them by adding selection, so user can select the lines they want to trim - why not?
For achieving this goal, we have to define how define the selected lines from the random selection (no expanded selection).
My question: is it worth? The current implementation simple to be maintained and working, I'm not sure a much more complicated solution will be accepted, especially the gain is not huge.

Trim operations need to be redefined and refined (or not, for both), this may take time. You can always do a feature request for it, though I don't promise you that the PR will be in the master.
From my side, I want to fix v8.4.8 RC to have regression/bug as less as possible. As a result, I have to revert https://github.com/notepad-plus-plus/notepad-plus-plus/commit/bf34ef0dd568b792627813a4445adf649a5c5ba6 - hope you understand.
